Fix nll_backward for negative weights.

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64275

Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

lezcano committed:
Fix nll_backward for negative weights.

Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

Lezcano commented:
cc @pmeier

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

[ghstack-poisoned]

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5271,6 +5271,7 @@ def sample_inputs_nll_loss(op_info, device, dtype, requires_grad, **kwargs):
         ((*shape, 1, 2), dict()),
         ((*shape, 1, 2, 3), dict()),
         (shape, dict(weight=make_tensor((num_classes,), device=device, dtype=dtype).abs())),
+        (shape, dict(weight=make_tensor((num_classes,), device=device, dtype=dtype))),
```
If we explicitly want to test this for positive and negative weights, maybe we should simply set `low=0` (and remove the `.abs()`) in the line above and here set 

```suggestion
        (shape, dict(weight=make_tensor((num_classes,), high=0, device=device, dtype=dtype))),
```

Lezcano replied:
Good point. Corrected!

pmeier replied:
Instead of leaving the `.abs()` call in the line above, could you just set `low=0` in the `make_tensor()` call? Or remove it completely to indicate "any kind of weight".

Lezcano replied:
Left one with positive, one with negative and one with mixed.

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

[ghstack-poisoned]

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

[ghstack-poisoned]

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

[ghstack-poisoned]

jbschlosser reviewed (changes requested):
Thanks @Lezcano for the fix! It looks like Jacobian checks are still failing on CUDA - guess we need similar changes in `native/cuda/Loss.cu` as well

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

Lezcano commented:
You are completely right @jbschlosser ! It should be fixed now.

I also removed some guards that did not give the right result when the sum of the weights was zero. If the sum of the passed weights is zero and the reduction is `mean`, this operation should return infinities, and same for the gradients. Otherwise it's silently returning a finite but incorrect result.

Lezcano commented:
Oh, sorry, the CI's failing. I'll fix it on Monday first thing.

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

Lezcano commented:
I believe this should be ready for review @jbschlosser

Lezcano commented:
Actually, I've realised that this is a linear function, and as such, the derivative should be implemented in terms of the forward (modulo saving some precomputations from the forward). I'm going to scrap what I have and factorize the code so that the forward and the backward use basically the same implementation.

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]

lezcano committed:
Update on "Fix nll_backward for negative weights."


Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

[ghstack-poisoned]
